firebase/firestore.rules

```diff
@@ -1,8 +1,164 @@
 rules_version = '2';
+
+function isAuthenticated() {
+  return request.auth.uid != null;
+}
+
+function isUser(id) {
+  return isAuthenticated() && request.auth.uid == id;
+}
+
+function isVerifiedCAV() {
+  // TODO: Actually use auth claims
+  return isAuthenticated() && true;
+}
+
+function isVerifiedPIN() {
+  // TODO: Actually use auth claims
+  return isAuthenticated() && true;
+}
+
+function currentUserIsCurrentPINRef(userID, database, data) {
+  return isAuthenticated() &&
+  ('pinUserRef' in data) &&
+  data.pinUserRef == /databases/$(database)/documents/users/$(userID)
+}
+
+function currentUserIsCurrentCAVRef(userID, database, data) {
+  return isAuthenticated() &&
+  ('cavUserRef' in data) &&
+  data.cavUserRef == /databases/$(database)/documents/users/$(userID)
+}
+
+function questionnaireBelongsToUser(userID, database, data) {
+    return isAuthenticated() &&
+           ('parentRef' in data) &&
+           data.parentRef == /databases/$(database)/documents/users/$(userID)
+}
+
+function confirmEitherPINorCAV(userID, database, data) {
+    return currentUserIsCurrentPINRef(userID, database, data) || currentUserIsCurrentCAVRef(userID, database, data)
+}
+
+function isNotSameCAVPIN(data) {
+  return data.cavUserRef != data.pinUserRef
```

Comment:
Is this situation possible? I feel like this case should be prevented upstream in the workflow by validating that a CAV can't make an offer on its own requests.

Reply:
Yes, but again we can't assume the clients writing data will always validate everything. They can have a bug.

Reply:
My bad, looking twice at the usages of the function, it's used exactly for what I meant!